File: mainline-sdk/typescript/src/tapSigner.ts
// Consumer-side TAP v2 receipt signing. Mirrors `mainline-sdk/rust/src/tap_signer.rs`
// and `mainline-service/src/billing/tap.rs` byte-for-byte.

import { secp256k1 } from "@noble/curves/secp256k1";
import { keccak_256 } from "@noble/hashes/sha3";

export interface TapDomain {
  settlementChainId: bigint;
  /** 20-byte GraphTallyCollector address. */
  verifyingContract: Uint8Array;
}

export interface TapReceiptV2 {
  /** 20-byte allocation id. */
  allocationId: Uint8Array;
  timestampNs: bigint;
  nonce: bigint;
  /** GRT wei (u128). */
  value: bigint;
  /** 65-byte (r || s || v) once signed. */
  signature: Uint8Array;
}

const enc = new TextEncoder();
const TAP_DOMAIN_TYPEHASH = enc.encode(
  "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)",
);
const TAP_RECEIPT_TYPEHASH = enc.encode(
  "Receipt(address allocation_id,uint64 timestamp_ns,uint64 nonce,uint128 value)",
);
const TAP_DOMAIN_NAME = enc.encode("TAP");
const TAP_DOMAIN_VERSION = enc.encode("2");

function keccak(b: Uint8Array): Uint8Array {
  return keccak_256(b);
}

function u256BE(value: bigint): Uint8Array {
  const out = new Uint8Array(32);
  let v = value;
  for (let i = 31; i >= 0 && v > 0n; i--) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

function concat(parts: Uint8Array[]): Uint8Array {
  const len = parts.reduce((a, p) => a + p.length, 0);
  const out = new Uint8Array(len);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

function addressWord(addr: Uint8Array): Uint8Array {
  const out = new Uint8Array(32);
  out.set(addr, 12);
  return out;
}

export function domainSeparator(d: TapDomain): Uint8Array {
  return keccak(
    concat([
      keccak(TAP_DOMAIN_TYPEHASH),
      keccak(TAP_DOMAIN_NAME),
      keccak(TAP_DOMAIN_VERSION),
      u256BE(d.settlementChainId),
      addressWord(d.verifyingContract),
    ]),
  );
}

export function structHash(r: TapReceiptV2): Uint8Array {
  return keccak(
    concat([
      keccak(TAP_RECEIPT_TYPEHASH),
      addressWord(r.allocationId),
      u256BE(r.timestampNs),
      u256BE(r.nonce),
      u256BE(r.value),
    ]),
  );
}

export function digest(d: TapDomain, r: TapReceiptV2): Uint8Array {
  return keccak(
    concat([new Uint8Array([0x19, 0x01]), domainSeparator(d), structHash(r)]),
  );
}

/**
 * Sign a receipt in place. Returns the 65-byte (r||s||v) signature with v in
 * legacy Ethereum form (27/28).
 */
export function sign(
  domain: TapDomain,
  receipt: TapReceiptV2,
  senderKey: Uint8Array,
): Uint8Array {
  if (senderKey.length !== 32) {
    throw new Error("sender key must be 32 bytes");
  }
  const d = digest(domain, receipt);
  const sig = secp256k1.sign(d, senderKey, { lowS: true, prehash: false });
  const out = new Uint8Array(65);
  const r = sig.toCompactRawBytes();
  out.set(r, 0);
  out[64] = (sig.recovery ?? 0) + 27;
  receipt.signature = out;
  return out;
}

export const RECEIPT_WIRE_VERSION = 1;
export const RECEIPT_WIRE_LEN = 1 + 20 + 8 + 8 + 16 + 65;

export function encodeWire(r: TapReceiptV2): Uint8Array {
  const out = new Uint8Array(RECEIPT_WIRE_LEN);
  let off = 0;
  out[off++] = RECEIPT_WIRE_VERSION;
  out.set(r.allocationId, off);
  off += 20;
  out.set(u64BE(r.timestampNs), off);
  off += 8;
  out.set(u64BE(r.nonce), off);
  off += 8;
  out.set(u128BE(r.value), off);
  off += 16;
  const sig = r.signature.length === 65 ? r.signature : padTo65(r.signature);
  out.set(sig, off);
  return out;
}

function u64BE(v: bigint): Uint8Array {
  const out = new Uint8Array(8);
  let x = v;
  for (let i = 7; i >= 0 && x > 0n; i--) {
    out[i] = Number(x & 0xffn);
    x >>= 8n;
  }
  return out;
}

function u128BE(v: bigint): Uint8Array {
  const out = new Uint8Array(16);
  let x = v;
  for (let i = 15; i >= 0 && x > 0n; i--) {
    out[i] = Number(x & 0xffn);
    x >>= 8n;
  }
  return out;
}

function padTo65(b: Uint8Array): Uint8Array {
  const out = new Uint8Array(65);
  out.set(b.slice(0, 65), 0);
  return out;
}

export function bytesToHex(b: Uint8Array): string {
  let s = "";
  for (const byte of b) s += byte.toString(16).padStart(2, "0");
  return s;
}

export function encodeHeader(r: TapReceiptV2): string {
  return bytesToHex(encodeWire(r));
}

File: mainline-sdk/rust/src/tap_signer.rs
//! Consumer-side TAP v2 receipt signing.
//!
//! Mirrors the typehashes and digest computation in
//! `mainline-service::billing::tap` so receipts produced here verify against
//! the operator with no further coordination.

use k256::ecdsa::{RecoveryId, Signature, SigningKey};
use sha3::{Digest, Keccak256};

#[derive(thiserror::Error, Debug)]
pub enum SignerError {
    #[error("invalid signing key")]
    InvalidKey,
    #[error("signing failed: {0}")]
    SigningFailed(String),
}

/// EIP-712 domain. Settlement-chain (Arbitrum One / Sepolia for Phase 0)
/// plus the deployed GraphTallyCollector address.
#[derive(Clone, Copy, Debug)]
pub struct TapDomain {
    pub settlement_chain_id: u64,
    pub verifying_contract: [u8; 20],
}

#[derive(Clone, Debug)]
pub struct TapReceiptV2 {
    pub allocation_id: [u8; 20],
    pub timestamp_ns: u64,
    pub nonce: u64,
    pub value: u128,
    /// Populated by [`sign`]; empty before signing.
    pub signature: Vec<u8>,
}

// Typehashes — must stay byte-identical to mainline-service::billing::tap.
const TAP_DOMAIN_TYPEHASH: &[u8] =
    b"EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)";
const TAP_RECEIPT_TYPEHASH: &[u8] =
    b"Receipt(address allocation_id,uint64 timestamp_ns,uint64 nonce,uint128 value)";
const TAP_DOMAIN_NAME: &[u8] = b"TAP";
const TAP_DOMAIN_VERSION: &[u8] = b"2";

fn keccak256(b: &[u8]) -> [u8; 32] {
    let mut h = Keccak256::new();
    h.update(b);
    h.finalize().into()
}

pub fn domain_separator(d: &TapDomain) -> [u8; 32] {
    let mut buf = Vec::with_capacity(32 * 5);
    buf.extend_from_slice(&keccak256(TAP_DOMAIN_TYPEHASH));
    buf.extend_from_slice(&keccak256(TAP_DOMAIN_NAME));
    buf.extend_from_slice(&keccak256(TAP_DOMAIN_VERSION));
    let mut chain = [0u8; 32];
    chain[24..].copy_from_slice(&d.settlement_chain_id.to_be_bytes());
    buf.extend_from_slice(&chain);
    let mut addr = [0u8; 32];
    addr[12..].copy_from_slice(&d.verifying_contract);
    buf.extend_from_slice(&addr);
    keccak256(&buf)
}

pub fn struct_hash(r: &TapReceiptV2) -> [u8; 32] {
    let mut buf = Vec::with_capacity(32 * 5);
    buf.extend_from_slice(&keccak256(TAP_RECEIPT_TYPEHASH));
    let mut alloc = [0u8; 32];
    alloc[12..].copy_from_slice(&r.allocation_id);
    buf.extend_from_slice(&alloc);
    let mut ts = [0u8; 32];
    ts[24..].copy_from_slice(&r.timestamp_ns.to_be_bytes());
    buf.extend_from_slice(&ts);
    let mut nonce = [0u8; 32];
    nonce[24..].copy_from_slice(&r.nonce.to_be_bytes());
    buf.extend_from_slice(&nonce);
    let mut val = [0u8; 32];
    val[16..].copy_from_slice(&r.value.to_be_bytes());
    buf.extend_from_slice(&val);
    keccak256(&buf)
}

pub fn digest(d: &TapDomain, r: &TapReceiptV2) -> [u8; 32] {
    let mut buf = Vec::with_capacity(2 + 32 + 32);
    buf.extend_from_slice(&[0x19, 0x01]);
    buf.extend_from_slice(&domain_separator(d));
    buf.extend_from_slice(&struct_hash(r));
    keccak256(&buf)
}

/// Sign a receipt in place. Returns the 65-byte (r || s || v) signature with
/// `v` in legacy Ethereum form (27/28).
pub fn sign(
    domain: &TapDomain,
    receipt: &mut TapReceiptV2,
    sender_key: &[u8; 32],
) -> Result<Vec<u8>, SignerError> {
    use k256::ecdsa::signature::hazmat::PrehashSigner;
    let key = SigningKey::from_bytes(sender_key.into()).map_err(|_| SignerError::InvalidKey)?;
    let d = digest(domain, receipt);
    let (sig, rec_id): (Signature, RecoveryId) =
        key.sign_prehash(&d).map_err(|e| SignerError::SigningFailed(e.to_string()))?;
    let mut out = Vec::with_capacity(65);
    out.extend_from_slice(&sig.to_bytes());
    out.push(rec_id.to_byte() + 27);
    receipt.signature = out.clone();
    Ok(out)
}

/// Wire format mirroring `mainline-service::billing::tap::encode_receipt`.
pub const RECEIPT_WIRE_VERSION: u8 = 1;
pub const RECEIPT_WIRE_LEN: usize = 1 + 20 + 8 + 8 + 16 + 65;

pub fn encode_wire(r: &TapReceiptV2) -> Vec<u8> {
    let mut out = Vec::with_capacity(RECEIPT_WIRE_LEN);
    out.push(RECEIPT_WIRE_VERSION);
    out.extend_from_slice(&r.allocation_id);
    out.extend_from_slice(&r.timestamp_ns.to_be_bytes());
    out.extend_from_slice(&r.nonce.to_be_bytes());
    out.extend_from_slice(&r.value.to_be_bytes());
    let mut sig = [0u8; 65];
    let n = r.signature.len().min(65);
    sig[..n].copy_from_slice(&r.signature[..n]);
    out.extend_from_slice(&sig);
    out
}

/// Convenience: produce the value to set on the `x-tap-receipt` gRPC
/// metadata header — hex-encoded wire form.
pub fn encode_header(receipt: &TapReceiptV2) -> String {
    hex::encode(encode_wire(receipt))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sign_emits_65_byte_v27_or_v28_signature() {
        let d = TapDomain { settlement_chain_id: 42161, verifying_contract: [0xcc; 20] };
        let mut r = TapReceiptV2 {
            allocation_id: [0xaa; 20],
            timestamp_ns: 1,
            nonce: 1,
            value: 1,
            signature: vec![],
        };
        let key = [0x11u8; 32];
        let sig = sign(&d, &mut r, &key).expect("sign");
        assert_eq!(sig.len(), 65);
        assert!(sig[64] == 27 || sig[64] == 28);
        assert_eq!(r.signature, sig);
    }

    #[test]
    fn header_encoding_is_hex_of_wire() {
        let d = TapDomain { settlement_chain_id: 42161, verifying_contract: [0xcc; 20] };
        let mut r = TapReceiptV2 {
            allocation_id: [0xaa; 20],
            timestamp_ns: 1,
            nonce: 1,
            value: 1,
            signature: vec![],
        };
        sign(&d, &mut r, &[0x11; 32]).unwrap();
        let header = encode_header(&r);
        assert_eq!(header.len(), RECEIPT_WIRE_LEN * 2);
        assert!(header.chars().all(|c| c.is_ascii_hexdigit()));
    }
}
